File: practice/src/main/java/com/decade/practice/persistence/redis/TypeEvent.java
package com.decade.practice.persistence.redis;

import com.decade.practice.persistence.jpa.embeddables.ChatIdentifier;
import com.decade.practice.utils.ChatUtils;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.redis.core.RedisHash;

import java.time.Instant;
import java.util.UUID;

@Data
@NoArgsConstructor
@RedisHash(value = "type", timeToLive = 2)
public class TypeEvent {

    private UUID from;
    private ChatIdentifier chat;
    // TODO: Adjust client
    private Instant time = Instant.now();

    @Id
    private String key;

    /**
     * Determines a key based on the from UUID and chat identifier.
     * This is a utility method that was a top-level function in the Kotlin version.
     */
    public static String determineKey(UUID from, ChatIdentifier chat) {
        UUID partner = ChatUtils.inspectPartner(chat, from);
        return from + "->" + partner;
    }

}

File: practice/src/main/java/com/decade/practice/model/TypeEvent.kt
package com.decade.practice.model

import com.decade.practice.model.embeddable.ChatIdentifier
import com.decade.practice.model.entity.Chat
import com.decade.practice.model.entity.User
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import java.util.*

fun determineKey(from: UUID, chat: ChatIdentifier): String {
    val partner = if (from == chat.firstUser)
        chat.secondUser else chat.firstUser
    return "$from->$partner"
}


@JsonDeserialize
@JsonSerialize
data class TypeEvent(
    val from: UUID,
    val chat: ChatIdentifier,
    val time: Long = System.currentTimeMillis(),
) {
    val key: String = determineKey(from, chat)

    constructor(owner: User, chat: Chat) : this(owner.id, chat.identifier)
}
